### ISSUE TITLE
Make current page label available as a regex value

### ISSUE THREAD

**Issue by charliehoward4dp** [DistributedProofreaders/guiguts#956]
When making targets for links in the HTML version, It sometimes is useful to use the current page number ("page label", not "image number") in a regex, but it often isn't easy to do so, because the "pagenum" may occur before or after the place it's needed.  Therefore, it would be helpful if Guiguts assigned the current page number to a reserved name group (if that's the correct terminology), e.g., $page or $pagenum. It would be used in the Replacement string. If there's no pagenum, the value could be set to something the user can find easily, e.g., "999999" or something better than that.

**Comment by windymilla** [comment]
Can you give or show an example please. Not sure what this means

**Comment by charliehoward4dp** [comment]
Given:
```
<div class="chapter">
<p><span class="pagenum" id="Page_3">3</span></p>

<h2 class="nobreak" id="I">I<br />

<span class="subhead"><i>Washington Irving</i></span></h2>
</div>
```
(where "pagenum" appears after "chapter")
or:

```
<p><span class="pagenum" id="Page_65">65</span></p>


<hr class="chap x-ebookmaker-drop" />

<div class="chapter">
<h2 class="nobreak" id="III">III<br />

<span class="subhead"><i>James Fenimore Cooper</i></span></h2>
</div>
```
(where "pagenum" appears before "chapter")

I'd like to add an anchor, id="toclink_3" or "id="toclink_65" in these examples. without having to look for the appropriate id="Page_(\d+?)", as it may occur before or after class="chapter">\n<h2, and there may be an arbitrary number of blank lines scattered through the area. So, this regex, using the proposed named variable $pagenum:

```
S: <div class="chapter">
R: <div id="toclink_$pagenum" class="chapter">
```

Would give this result, whether `<span class="pagenum"...>` is before or after the `<div class="chapter">`:
```
<div id="toclink_3" class="chapter">
<p><span class="pagenum" id="Page_3">3</span></p>

<h2 class="nobreak" id="I">I<br />

<span class="subhead"><i>Washington Irving</i></span></h2>
</div>
```

or:
```
<p><span class="pagenum" id="Page_65">65</span></p>


<hr class="chap x-ebookmaker-drop" />

<div id="toclink_65" class="chapter">
<h2 class="nobreak" id="III">III<br />

<span class="subhead"><i>James Fenimore Cooper</i></span></h2>
</div>
```

Examples of other situations that could use the variable involve `<h3>` or illustrations.  Currently, we make do by linking to the page or a nearby page, but this would let us link directly to the specific item being referenced.

**Comment by windymilla** [comment]
Do I detect [history repeating itself](https://www.pgdp.net/phpBB3/viewtopic.php?p=948657#p948657) (plus the 13 subsequent posts)?

The method there puts an id based on the "current" pagenum (i.e. the page number when the h2 is encountered)

Here, you seem to be asking the regex engine to make a judgement as to whether to use the current page number (i.e. the number set before the h2 appears), or to look ahead to see if a page number is about to be set, maybe after some blank lines, and then use that. I think that is beyond the reasonable expectations for a regex. I'm not saying impossible, but definitely harder than the above method, which many years ago you described as "clever^2" :)

For consistency of display, I would recommend ensuring your page number always precedes your h2. Also, this should make the above method work for you. If you agree, please feel free to close this. If I've misunderstood, or you have an idea of how to make it work, just let me know and I'll see what I can do.

**Comment by charliehoward4dp** [comment]
I want the number shown on the Status bar’s “Lbl:” box assigned to $pagenum, even if it’s a Roman number; but without the “Pg “. If there’s no number, “999999” will do, as a signal.

GG apparently begins each page at the crlf preceding the first line of that page, even when the next line also is just a crlf.

If this doesn’t seem to be useful to others, I can continue using my two-pass regex method.

**Comment by windymilla** [comment]
Ah, that may be possible. I'll have a look

**Comment by windymilla** [comment]
OK. The problem is that the system only updates the Lbl label in the status bar to the correct value when the cursor is there, but it had nearly all the information and code needed - the "flexible" change described further below required just one additional line of code and a small change to one other line. The less flexible "short" version needed only 3 extra lines of code.

Short version: I've added `\P` (no `\E` needed) which I think gives you what you want - use it instead of `$pagenum` in your example. The `P` stands for "page label", since `\L` is already used for lowercase.

`<tb>`

Long version, which I wrote first, then realised it sounded a bit cumbersome, and it gave me the idea for the short version. 

I've gone a little off piste to take a slightly different, but more flexible approach than your direct request. 

I've added a new extended regex code `\I` (similar to `\C...\E`, but no `\E` needed because `\I` doesn't enclose a string).

This code gets replaced with the index of the current match in the file. In your example, this would be the line and column of the part of the `h2` element that your regex matches. 

You therefore need to use a replace string something like this:
`\C$::pagenumbers{"Pg".::get_page_number("\I")}{label}\E`

This means

1. Execute perl code
2. Access the pagenumbers hash table stored in the code
3. Get the page number of the index position of the regex match (using `\I`) and prefix "Pg" to it (using the dot operator)
4. Extract the "label" that corresponds to that page
5. Stop executing code

As I said, this adds flexibility for future requests that need the index position to get some other data, but perhaps it's an unnecessary complexity - this is the point I thought of the short version described above.

If you've managed to read this far, I'd value your opinion on whether both `\I` and `\P` work in a useful way. Note that you can use `\I` on its own, without all the other code around it and see what it gives you - should be "row.column". You could also use `\C::get_page_number("\I")\E` to get the page number rather than the page label. Any comments on whether to retain this feature or remove it and just leave the short `\P` also gratefully received.

(Carefully placed at the end to force you to read down to here - haha) To try this out you'll need the two files in [pagelabel.zip](https://github.com/DistributedProofreaders/guiguts/files/9077915/pagelabel.zip).

**Comment by charliehoward4dp** [comment]
I would have, and did, read all of it anyway, because I was trying to figure out what you were getting to, and whether I would have to use something fancy rather than simply \P. I did appreciate your ha ha, especially since I wasn’t expecting something to test so quickly. Am sending this message now since I know that it is late where you are. Am about to download and try those patches, and will let you know what I find. I will make use of\P, will test \I, and as to the rest, we shall see. Thank you for doing so much so soon.

**Comment by wf49670** [comment]
Isn't \P already in use, with a slightly different syntax? At least in Perl regex syntax there's \P{} with a named property between the braces.

**Comment by charliehoward4dp** [comment]
First results, only for \P: it works fine, but it would be more useful if only the number and not the "Pg " (with a space) was assigned to \P, because its purpose (well, my purpose in asking for it) is to be the target of a link, and links don't mix with spaces.  Other than that, it's really terrific, because it made it trivially easy to link ToC section page numbers directly to the `<h3>`'s, rather than to the tops of the pages they are (supposed to be) on.  The "Pg" isn't useful, even though that's what's in the "Lbl" field, because the books themselves just use page numbers.

**Comment by charliehoward4dp** [comment]
When there's no page label, `\P` is null; it would be more useful for it to contain something like "99999", to avoid the risk of generating syntactically invalid links. (The user still will have to search for that placeholder, as there likely will be duplicates.)  However, this is not an essential change, as user cleanup will be needed afterwards either way. It is essential, IMO, to omit the "Pg " (with a space), or at least, to omit the space, to avoid a syntax error.

`\I` by itself generates just what you said it should, and so does `\C$::pagenumbers{"Pg".::get_page_number("\I")}{label}\E`. Currently, I don't know how I'd use them, as the text of the books use page numbers, but I suspect you have some other purposes in mind, that will reveal themselves in their own good time.

**Comment by windymilla** [comment]
Thanks for the feedback, @charliehoward4dp - I have a version ready to merge in now. There's a copy of the relevant files for you in [PageLabel2.zip](https://github.com/DistributedProofreaders/guiguts/files/9079404/PageLabel2.zip) while you wait for the new release.

I have removed `\I` completely, and have used `\IP` to signify "Insert Page label". This leaves flexibility for inserting other things in the future using `\I` plus another letter. I could envisage you wanting to insert the line number or column number to feature in a math expression involving something to do with table markup, or maybe the current chapter number, or who knows what else?

I also stripped the leading `Pg ` and used `999999` as a replacement for empty "No count" labels as originally requested.
